aws-api/src/main/java/com/amplifyframework/api/aws/AppSyncGraphQLRequest.java

```diff
@@ -0,0 +1,295 @@
+ */
+public final class AppSyncGraphQLRequest<R> extends GraphQLRequest<R> {
+    private static final int DEFAULT_LEVEL_DEPTH = 2;
+
+    private final ModelSchema modelSchema;
+    private final OperationType operationType;
+    private final SelectionSet.Node selectionSet;
+    private final Map<String, Object> variables;
+    private final Map<String, String> variableTypes;
+
+    /**
+     * Constructor for AppSyncGraphQLRequest.
+     * @throw AmplifyException if a ModelSchema can't be derived from the provided model class
+     */
+    private AppSyncGraphQLRequest(Builder builder) throws AmplifyException {
+        super(builder.responseType, new GsonVariablesSerializer());
+        this.modelSchema = ModelSchema.fromModelClass(builder.modelClass);
+        SelectionSet.Node set = SelectionSet.fromModelClass(builder.modelClass, DEFAULT_LEVEL_DEPTH);
+        if (QueryType.LIST.equals(builder.operationType)) {
+            set = SelectionSet.wrapPagination(set);
+        }
+        this.selectionSet = set;
+        this.operationType = builder.operationType;
+        this.variables = builder.variables;
+        this.variableTypes = builder.variableTypes;
+    }
+
+    /**
+     * Copy constructor for an AppSyncGraphQLRequest.
+     * @param request request to copy.
+     * @param <R> response type.
+     */
+    public <R> AppSyncGraphQLRequest(AppSyncGraphQLRequest<R> request) {
+        super(request);
+        this.modelSchema = request.modelSchema;
+        this.operationType = request.operationType;
+        this.selectionSet = new SelectionSet.Node(request.selectionSet);
+        this.variables = new HashMap<>(request.variables);
+        this.variableTypes = new HashMap<>(request.variableTypes);
+    }
+
+    @Override
+    public <R> AppSyncGraphQLRequest<R> copy() {
+        return new AppSyncGraphQLRequest<R>(this);
+    }
+
+    @Override
+    public Map<String, Object> getVariables() {
+        return this.variables;
+    }
+
+    /**
+     * Used for setting a variable on the request.
+     * @param key variable name
+     * @param type type of variable value
+     * @param value variable value
+     */
+    public void setVariable(String key, String type, String value) {
+        this.variables.put(key, value);
+        this.variableTypes.put(key, type);
+    }
+
+    /**
+     * Used to add owner argument from authProvider if needed.
+     * @param authProvider CognitoUserPoolsAuthProvider for obtaining the username to set as the owner field.
+     * @throws ApiException if request requires owner argument and authProvider or authProvider.getUsername() is null.
+     */
+    public void setAuthProvider(CognitoUserPoolsAuthProvider authProvider) throws ApiException {
+        for (AuthRule authRule : modelSchema.getAuthRules()) {
+            if (isOwnerArgumentRequired(authRule.getOperationsOrDefault())) {
+                setVariable(authRule.getOwnerFieldOrDefault(), "String!", getUsername(authProvider));
```

Comment:
So I can't explain why we have to set `owner: String!` and the value on graphQL document/variables. We already intercept the request with the token which contains all of this information, so shouldn't the service already know how to retrieve the token and check if the user is authorized? Why do both? @SwaySway tried to explain this to me once and it had something to do with how subscriptions work. There is the initial connection that contains the auth token on the request header and then on each subscription message, it probably requires the `owner` value to be used as a filter to check if the data can be passed back to the subscriber?

So from trying out different auth rules with "allow: owner", it turns out, as long as the operation is specified then the correlating operation's subscription requires the owner value to be present. This is what you have written here. The callout here is that this does not work for all auth rules. I don't think this should block this PR from shipping as this is just a limitation that we can document and increase in functionality later. 

The problem is when there are auth rules with "allow: groups, groups: ["SomeGroup"]". 
```
type ApprovedBlog @model @auth(rules: [{allow: groups, groups: ["ApprovedBloggers"], operations: [read, create]}]) {
  id: ID!
  blogId: String
}
```
ApprovedBlog can only be read and created by users in the "ApprovedBloggers" group. (Uh, i'm not sure what this means for others not in the group, maybe they can update and delete ApprovedBlog objects if they have access to the id of the object? not sure)

None of the generated subscription operations require the owner field
```
onCreateApprovedBlog: ApprovedBlog
onUpdateApprovedBlog: ApprovedBlog
onDeleteApprovedBlog: ApprovedBlog
```
I have `user1` not in the group and `user2` in the group.

Both logged in users create a subscription without the owner:
```
subscription onCreaeApprovedBlog {
	onCreateApprovedBlog {
    id
    blogId
  }  
}
```

user1 fails with 
```
Error: {
    "errors": [
        {
            "message": "Connection failed: {\"errors\":[{\"errorType\":\"Unauthorized\",\"message\":\"Not Authorized to access onCreateApprovedBlog on type Subscription\"}]}"
        }
    ]
}
```
user2 is able to create a subscription connection.

Since models with static groups auth rules do not require the `owner` to be passed in, the check for `isOwnerArgumentRequired` should be: authRule's allow == owner && `SubscriptionType` match `ModelOperation`

Reply:
> Since models with static groups auth rules do not require the `owner` to be passed in, the check for isOwnerArgumentRequired should be: authRule's allow == owner && `SubscriptionType` match `ModelOperation`

Thanks for this clarification, I just added a check to make sure the owner argument is only added for the `owner` strategy.

> there are other related [issues](https://github.com/aws-amplify/amplify-ios/issues/576) to keep in mind which may or may not be relevant on Android related to having both "allow: owner" and "allow: groups"

I will keep this in mind, as my next task is to incorporate the auth directive into `DataStore`.